File: 2019/LevelUpDecBatch/l007_HashMapAndHeap/M001_HeapConstruction/heapSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

bool compareTo(vector<int> &arr, bool isIncreasing, int i, int j)
{
    if (isIncreasing)
        return arr[i] > arr[j];
    else
        return arr[i] < arr[j];
}

void downHepify(vector<int> &arr, bool isIncreasing, int pi, int li)
{
    int maxIdx = pi;
    int lci = 2 * pi + 1;
    int rci = 2 * pi + 2;

    if (lci <= li && compareTo(arr, isIncreasing, lci, maxIdx))
        maxIdx = lci;
    if (rci <= li && compareTo(arr, isIncreasing, rci, maxIdx))
        maxIdx = rci;

    if (pi != maxIdx)
    {
        swap(arr[pi], arr[maxIdx]);
        downHepify(arr, isIncreasing, maxIdx, li);
    }
}

void heapSort(vector<int> &arr, bool isIncreasing)
{
    int li = arr.size() - 1;
    for (int i = li; i >= 0; i--)
        downHepify(arr, isIncreasing, i, li);

    while (li >= 0)
    {
        swap(arr[0], arr[li--]);
        downHepify(arr, isIncreasing, 0, li);
    }
}

int main()
{
    vector<int> arr = {10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13};
    heapSort(arr, true);

    for (int ele : arr)
        cout << ele << " ";
    return 0;
}

File: 2020/pp-mar 14/lecture_011_HeapAndHashMap/HeapConstruction/heapSort.java
public class heapSort {

    public static int compareTo(int[] arr, int t, int o, boolean isIncreasing) {
        if (isIncreasing) {
            return arr[t] - arr[o];
        } else {
            return arr[o] - arr[t];
        }
    }

    public static int swap(int i, int j) {
        return i;
    }

    public static void downHeapify(int[] arr, int pi, int lidx, boolean isIncreasing) {
        int maxIdx = pi, lci = 2 * pi + 1, rci = 2 * pi + 2;
        if (lci <= lidx && compareTo(arr, lci, maxIdx, isIncreasing) > 0)
            maxIdx = lci;
        if (rci <= lidx && compareTo(arr, rci, maxIdx, isIncreasing) > 0)
            maxIdx = rci;

        if (pi != maxIdx) {
            arr[maxIdx] = swap(arr[pi], arr[pi] = arr[maxIdx]);
            downHeapify(arr, maxIdx, lidx, isIncreasing);
        }

    }

    public static void heapSort(int[] arr, boolean isIncreasing) {
        int n = arr.length, lidx = n - 1;
        for (int i = lidx; i >= 0; i--) {
            downHeapify(arr, i, lidx, isIncreasing);
        }

        while (lidx >= 0) {
            arr[lidx] = swap(arr[0], arr[0] = arr[lidx]);
            downHeapify(arr, 0, --lidx, isIncreasing);
        }

    }

    public static void display(int[] arr) {
        for (int ele : arr)
            System.out.print(ele + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = { 10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13, 14 };
        boolean isIncreasing = false;
        heapSort(arr, isIncreasing);

        display(arr);
    }
}